SNAPSHOT.md

```diff
@@ -0,0 +1,94 @@
+# Requirements of `AsyncContext.Snapshot`
+
+`AsyncContext.Snapshot` presents two unique requirements:
+
+- It does not expose the value associated with a `Variable` instance.
+- It captures _all_ `Variable`s' current value and restores those values
+  at a later time.
+
+The above requirements are essential to decouples a queueing implementation
+from the consumers of `Variable` instances. For example, a scheduler can queue
+an async task and taking a snapshot of the current context:
+
+```typescript
+// The scheduler doesn't access to any AsyncContext.Variable.
+const scheduler = {
+  queue: [],
+  postTask(task) {
+    // Each callback is stored with the context at which it was enqueued.
+    const snapshot = new AsyncContext.Snapshot();
+    queue.push(() => snapshot.run(task));
```

Comment:
Nit: we actually have 2 callbacks (`task`, and the anonymous arrow) here, and that makes the explanation in `runWhenIdle` a little confusing. Can we store the `snapshot` and `task` in the queue, then call `snapshot.run(task)` in our loop?